pkg/apis/admissionregistration/types.go

```diff
@@ -115,6 +115,244 @@ const (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// ValidatingAdmissionPolicy describes the definition of an admission validation policy that accepts or rejects and object without changing it.
+type ValidatingAdmissionPolicy struct {
+	metav1.TypeMeta
+	// Standard object metadata; More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata.
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of the AdmissionPolicy.
+	Spec ValidatingAdmissionPolicySpec
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ValidatingAdmissionPolicyList is a list of ValidatingAdmissionPolicy.
+type ValidatingAdmissionPolicyList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta
+	// List of ValidatingAdmissionPolicy.
+	Items []ValidatingAdmissionPolicy
+}
+
+// ValidatingAdmissionPolicySpec is the specification of the desired behavior of the AdmissionPolicy.
+type ValidatingAdmissionPolicySpec struct {
+	//ParamSource specifies the kind of resources used to parameterize this policy which has to be a CRD.
+	// +optional
+	ParamSource *ParamSource
+
+	// MatchConstraints specifies what resources this policy is designed to validate.
+	// The AdmissionPolicy cares about a validation if it matches _any_ Constriant.
+	// However, in order to prevent clusters from being put into a unustable state that cannot be recoverd from via the API
+	// alidatingAdmissionPolicy cannot match ValidatingAdmissionPolicy/PolicyBinding/param resources.
+	// ValidatingWebhookConfiguration cannot match MutatingWebhookConfiguration or ValidatingAdmissionPolicy/PolicyBinding/param resources.
+	// +optional
+	MatchConstraints *MatchResources
+
+	// Validations contain CEL expressions which is used to apply the validation.
+	// Required.
+	Validations []Validation
+
+	// FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+	// allowed values are Ignore or Fail. Defaults to Fail.
+	// +optional
+	FailurePolicy *FailurePolicyType
+
+	// SingletonBinding is for the use case that a policy can be authored using a single ValidatingAdmissionPolicy resource.
+	// This is only available for cases where there is no need to have multiple bindings, and where all params can be inlined in CEL.
+	// If spec.singletonBinding is present policy binding support is disabled.
+	// spec.paramSource must be absent to use this feature.
+	// +optional
+	SingletonBinding *MatchResources
```

Comment:
Question: Could it be a bool and use MatchConstraints for matching?

Reply:
Need to have a separate struct so we can add other fields in addition to matchresources in the future?

Reply:
I agree that MatchConstraints is sufficient for all matching for singletons. Maybe SingletonBinding should have a type that is currently an empty struct? We will need fields there before alpha is done, so maybe it's not crazy?

cc @liggitt